tenders/common/elf.c

```diff
@@ -73,120 +75,237 @@ static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
+         * the Solo5 toolchains do not produce such executables.
          */
-        if (align > 0 && (align & (align - 1)) == 0) {
-            if (add_overflow(result, (align - 1), _end))
-                goto out_invalid;
-            _end = _end & -align;
-        }
-        else {
-            _end = result;
+        if (p_vaddr_start < e_end)
+            goto out_invalid;
+        /*
+         * Verify p_vaddr + p_filesz is within range.
+         */
+        if (p_vaddr >= mem_size)
+            goto out_invalid;
+        if (add_overflow(p_vaddr, p_filesz, temp))
+            goto out_invalid;
+        if (temp > mem_size)
+            goto out_invalid;
+        /*
+         * Compute p_vaddr_end = p_vaddr + p_memsz, aligned up to requested
+         * alignment and verify result is within range.
+         */
+        if (p_memsz < p_filesz)
+            goto out_invalid;
+        if (add_overflow(p_vaddr, p_memsz, p_vaddr_end))
+            goto out_invalid;
+        if (align_up(p_vaddr_end, p_align, &p_vaddr_end))
+            goto out_invalid;
+        if (p_vaddr_end > mem_size)
+            goto out_invalid;
+        /*
+         * Keep track of the highest byte of memory occupied by the program.
+         */
+        if (p_vaddr_end > e_end) {
+            e_end = p_vaddr_end;
+            /*
+             * Double check result for host (caller) address space overflow.
+             */
+            assert((mem + e_end) >= (mem + p_min_loadaddr));
         }
-        if (_end > *p_end)
-            *p_end = _end;
 
-        daddr = mem + paddr;
-        nbytes = pread_in_full(fd_kernel, daddr, filesz, offset);
+        /*
+         * Load the segment (p_vaddr .. p_vaddr + p_filesz) into host memory at
+         * host_vaddr and ensure any BSS (p_memsz - p_filesz) is initialised to
+         * zero.
+         */
+        uint8_t *host_vaddr = mem + p_vaddr;
+        /*
+         * Double check result for host (caller) address space overflow.
+         */
+        assert(host_vaddr >= (mem + p_min_loadaddr));
+        nbytes = pread_in_full(fd_kernel, host_vaddr, p_filesz,
+                phdr[ph_i].p_offset);
         if (nbytes < 0)
             goto out_error;
-        if (nbytes != filesz)
+        if (nbytes != p_filesz)
             goto out_invalid;
-        memset(daddr + filesz, 0, memsz - filesz);
+        memset(host_vaddr + p_filesz, 0, p_memsz - p_filesz);
 
-        prot = PROT_NONE;
+        /*
+         * Memory protection flags should be applied to the aligned address
+         * range (p_vaddr_start .. p_vaddr_end). Before we apply them, also
+         * verify that the address range is aligned to the architectural page
+         * size.
+         */
+        if (p_vaddr_start & (EM_PAGE_SIZE - 1))
+            goto out_invalid;
+        if (p_vaddr_end & (EM_PAGE_SIZE - 1))
+            goto out_invalid;
+        host_vaddr = mem + p_vaddr_start;
```

Comment:
I find the reuse of this variable (`host_vaddr`) a bit confusing.
In the blocks above, `host_vaddr = mem + p_vaddr` is being used to load `p_filesz` bytes (unaligned).
Then additional stuff after that is `memset` up until length `mem + p_memsz`.

In this line here it is reassigned as `host_vaddr = mem + p_vaddr_start` (aligned, rounded down), after which the mapping between `mem + v_addr_start` and `mem + align_up(p_memsz, p_align)` is `mprotect()'ed` using the desired flags.
What is the intended contents of `mem + align_down(p_vaddr, p_align)` through `mem + p_vaddr` at this point? Should that also be cleared, or is the caller expected to have cleared `mem ... mem+mem_size`?

Reply:
> I find the reuse of this variable (host_vaddr) a bit confusing.

Fair enough. How about I introduce a `host_vaddr_start` for the second block?

> What is the intended contents of mem + align_down(p_vaddr, p_align) through mem + p_vaddr at this point?

The [ELF specification](http://refspecs.linuxfoundation.org/elf/elfspec.pdf) discusses this on page 2-8:

> ‘‘Impurities’’ in the other three pages are not logically part of the process image; whether the system expunges them is unspecified.

I.e. we don't care. In practice, given that the host memory comes from `mmap()` it will on all modern systems be initialised to zero.